Loot table preview

BaR1BoD opened the pull request:
This PR adds a preview panel for the LootTable ME

![Screenshot 2025-05-21 223824](https://github.com/user-attachments/assets/8f41d0fb-bf2c-4917-b23e-64ead70136ca)

KlemenDEV commented:
Interesting idea! :D

KlemenDEV commented on plugins/mcreator-themes/themes/default_dark/images/container.png:
Alright, thanks for the confirmation! :)

KlemenDEV reviewed (commented):
In some cases, it looks like multiple numbers are generated for the same slot:

![image](https://github.com/user-attachments/assets/f68d3cba-27f8-4c82-a997-0061aae10495)

KlemenDEV commented:
It may also make sense to omit air entires from rendering

KlemenDEV commented:
> net.mcreator.ui.minecraft.loottable.LootTablePreview.generateLootTable(LootTablePreview.java:97)

It also seems some bound checking can underflow in the preview generator

KlemenDEV commented:
Ready for review?

KlemenDEV commented on src/main/java/net/mcreator/ui/modgui/LootTableGUI.java:
```diff
@@ -156,6 +163,8 @@ public LootTableGUI(MCreator mcreator, ModElement modElement, boolean editingMod
 
 		lootTablePools.setEntries(loottable.pools);
+
+		listener.registerUI(lootTablePools);
```
I know this is not ready for review yet, but I just wanted to let you know this is not ok that you only register the listener in editing mode. This should work in all cases, not just while editing. Why did you go for this option?

BaR1BoD replied:
`listener.registerUI()` is calling `modElementChanged()`, which will try to generate the preview, but as `openInEditingMode()` is being called after `initGUI()`, somehow the entries minimum count has been set but their maximum count hasn't been set (which is set to1 by default)

But if it is not okay, a check like this in `generateLootTable()` would work too
```java
if (entry.min > entry.max) {
    continue;
}
```

BaR1BoD commented:
> In some cases, it looks like multiple numbers are generated for the same slot:

I couldn't reproduce it. Could you please show me the loottable pools

KlemenDEV commented:
Here is an example:

![image](https://github.com/user-attachments/assets/a25552cc-c3c8-407b-b2d0-2afe9cdadad0)

(see strange 4 letters for the "blue sticks")

I have also noticed now with recent commits preview is no longer updated upon opening ME in edit mode immediately

BaR1BoD commented:
It seems that the for loop inside `ModElementChangedListener.registerUI()` is causing an infinite loop (It happens for the second test example). I have no idea why this is happening.

Goldorion commented on plugins/mcreator-themes/themes/default_dark/images/container.png:
By using `MinecraftImageGenerator#generateInventorySlots` with `MIG#generateBackground`, it's possible to remove the use of this (surely copyrighted) image.

KlemenDEV commented:
I will try to investigate the loop and potential solution asap :)

KlemenDEV commented:
RFR?

BaR1BoD commented:
I have fixed the duplicated numbers, tests failure, and also remove the container.png and make the container via `MIG#generateBackground` and `MIG#generateItemSlot`.

Additionally, I added an item splitter to address an issue where entries with a single roll but items with a count greater than 1 would stack all items in one slot. The item splitter now distributes them across separate slots.

AAli107 commented:
This should have a randomize button for the preview so that you can randomize the loot table preview to ensure the loot table is balanced.

KlemenDEV commented:
@BaR1BoD, the idea from @AAli107 is not bad. A button that would use random to generate a new random seed, that is then used until the button is clicked again.

This way preview does not jump with every loot table change as seed is constant, but seed can still be changed by the user

Herobrine7707 commented:
A preview for lootable would be cool

KlemenDEV reviewed (changes requested):
This loot table is always empty, no matter how many times I click new seed:

<img width="1556" height="506" alt="image" src="https://github.com/user-attachments/assets/50a6a946-a0ab-4e34-b25f-a16b472f7705" />

KlemenDEV reviewed (approved):
LGTM, thanks for this cool feature!
